consensus/src/miner.rs

```diff
@@ -97,106 +92,116 @@ impl<S: Storage> Miner<S> {
+                    .into_iter()
+                    .map(|x| x.into())
+                    .collect(),
+            )
+            .await?;
+
+        transactions.push(response.transaction);
+        Ok(response.records)
     }
 
     /// Acquires the storage lock and returns the previous block header and verified transactions.
     #[allow(clippy::type_complexity)]
-    pub fn establish_block(
+    pub async fn establish_block(
         &self,
-        transactions: &DPCTransactions<Tx>,
-    ) -> Result<(BlockHeader, DPCTransactions<Tx>, Vec<DPCRecord<Components>>), ConsensusError> {
-        let rng = &mut thread_rng();
-        let mut transactions = transactions.clone();
-        let coinbase_records = self.add_coinbase_transaction(&mut transactions, rng)?;
+        transactions: &[SerialTransaction],
+    ) -> Result<(Vec<SerialTransaction>, Vec<SerialRecord>), ConsensusError> {
+        let mut transactions = transactions.to_vec();
+        let coinbase_records = self.add_coinbase_transaction(&mut transactions).await?;
 
         // Verify transactions
-        assert!(InstantiatedDPC::verify_transactions(
-            &self.consensus.public_parameters,
-            &transactions.0,
-            &*self.consensus.ledger,
-        )?);
+        // assert!(InstantiatedDPC::verify_transactions(
+        //     &self.consensus.public_parameters,
+        //     &transactions,
 
-        let previous_block_header = self.consensus.ledger.get_latest_block()?.header;
+        // )?);
 
-        Ok((previous_block_header, transactions, coinbase_records))
+        Ok((transactions, coinbase_records))
     }
 
     /// Run proof of work to find block.
     /// Returns BlockHeader with nonce solution.
-    pub fn find_block<T: TransactionScheme>(
+    pub fn find_block(
         &self,
-        transactions: &DPCTransactions<T>,
-        parent_header: &BlockHeader,
-    ) -> Result<BlockHeader, ConsensusError> {
-        let txids = transactions.to_transaction_ids()?;
+        transactions: &[SerialTransaction],
+        parent_header: &SerialBlockHeader,
+    ) -> Result<SerialBlockHeader, ConsensusError> {
+        let txids = transactions.iter().map(|x| x.id).collect::<Vec<_>>();
         let (merkle_root_hash, pedersen_merkle_root_hash, subroots) = txids_to_roots(&txids);
 
         let time = Utc::now().timestamp();
         let difficulty_target = self.consensus.parameters.get_block_difficulty(parent_header, time);
 
         // TODO: Switch this to use a user-provided RNG
-        let (nonce, proof) = self.miner.mine(
+        let (nonce, proof) = POSW.mine(
             &subroots,
             difficulty_target,
             &mut thread_rng(),
             self.consensus.parameters.max_nonce,
         )?;
 
         Ok(BlockHeader {
-            previous_block_hash: parent_header.get_hash(),
+            previous_block_hash: BlockHeaderHash(parent_header.hash().bytes().unwrap()),
```

Comment:
Will this not fail even for the genesis block?

Reply:
the miner doesn't mine the genesis block. there is also always a parent_header contractually present here.